src/main/java/com/android/volley/toolbox/BasicAsyncNetwork.java

```diff
@@ -0,0 +1,419 @@
+                                        request,
+                                        requestStartMs,
+                                        mBaseHttpStack.executeRequest(
+                                                request, additionalRequestHeaders),
+                                        callback);
+                            } catch (AuthFailureError e) {
+                                callback.onError(e);
+                            } catch (IOException e) {
+                                onRequestFailed(request, callback, e, requestStartMs, null, null);
+                            }
+                        }
+                    });
+        }
+    }
+
+    /**
+     * This method sets the non blocking executor to be used by the stack for non-blocking tasks. If
+     * you are not using an {@link com.android.volley.toolbox.AsyncHttpStack}, this should do
+     * nothing. This method must be called before performing any requests if you are using an
+     * AsyncHttpStack.
+     */
+    @Override
+    public void setNonBlockingExecutorForStack(ExecutorService executor) {
+        if (mBaseHttpStack instanceof AsyncHttpStack) {
+            AsyncHttpStack stack = (AsyncHttpStack) mBaseHttpStack;
+            stack.setNonBlockingExecutor(executor);
+        }
+    }
+
+    /**
+     * This method sets the blocking executor to be used by the network and stack for potentially
+     * blocking tasks. This method must be called before performing any requests. Only set the
+     * executor for the stack if it is an instance of {@link
+     * com.android.volley.toolbox.AsyncHttpStack}
+     */
+    @Override
+    public void setBlockingExecutor(ExecutorService executor) {
+        mBlockingExecutor = executor;
+        if (mBaseHttpStack instanceof AsyncHttpStack) {
+            AsyncHttpStack stack = (AsyncHttpStack) mBaseHttpStack;
+            stack.setNonBlockingExecutor(executor);
+        }
+    }
+
+    /**
+     * Attempts to prepare the request for a retry. If there are no more attempts remaining in the
+     * request's retry policy, a timeout exception is thrown.
+     *
+     * @param request The request to use.
+     */
+    private void attemptRetryOnException(
+            final String logPrefix,
+            final Request<?> request,
+            final OnRequestComplete callback,
+            final VolleyError exception) {
+        final RetryPolicy retryPolicy = request.getRetryPolicy();
+        final int oldTimeout = request.getTimeoutMs();
+
+        try {
+            retryPolicy.retry(exception);
+        } catch (VolleyError e) {
+            request.addMarker(
+                    String.format("%s-timeout-giveup [timeout=%s]", logPrefix, oldTimeout));
+            callback.onError(e);
+            return;
+        }
+        request.addMarker(String.format("%s-retry [timeout=%s]", logPrefix, oldTimeout));
+        // If we are using an async stack, then perform retry after a timeout.
```

Comment:
I had also considered having the network user set the timeout for retrying again. Then we could also check which stack they're using there instead of in this method, and just default to 0 if non-async stack. Wasn't sure though about this.

Reply:
It looks like what the current code is doing is just trying again after an unconfigurable 100ms delay, which doesn't seem right to me.

What I think we should do for now is match the existing behavior as much as possible with our current APIs. That would mean retrying immediately upon failure. In theory, when the request is retried, getTimeoutMs will return an increased time unit when passed to Cronet. However, as we established earlier, Cronet doesn't actually support setting a timeout, so this increased timeout gets ignored. But that's an implementation detail of the HTTP stack, and other stacks might respect the timeout.

I do think we should make it possible for clients to configure a gap between retries, as the doc noted. But this would be a separate concept which requires a new API, and I don't think we should assume a fixed 100ms retry interval for each retry.